### ISSUE TITLE
rails6.1.1 counter_cache conflict lock_version when destroy a object

### ISSUE THREAD

**Issue by Guoxweii** [rails/rails#41655]
### System configuration
**Rails version**:  6.1.1

**Ruby version**: 2.7.2

#### Example

~~~
class Post
  has_many :comments, dependent: :destroy, counter_cache: "comments_count"
end

class Comment
  lock_version # has a lock_version column
  belongs_to :post, counter_cache: "comments_count", touch: true
end

~~~

**when i destroy a Comment object, the relate Post object's counter_cache won't trigger.**

Debug result:

when destroy a Comment object, 
 
path: activerecord-6.1.1/lib/active_record/locking/optimistic.rb

~~~
     def destroy_row
          return super unless locking_enabled?

          locking_column = self.class.locking_column

          affected_rows = self.class._delete_record(
            @primary_key => id_in_database,
            locking_column => attribute_before_type_cast(locking_column)
          )

          if affected_rows != 1
            raise ActiveRecord::StaleObjectError.new(self, "destroy")
          end

          affected_rows
        end
~~~

because Comment model has a lock_version, so the locking_enabled? is true

~~~ ruby
return super unless locking_enabled?
~~~

so super(CounterCache's destroy_row function) not trigger, and the count_cache decrease not trigger.

how can l resolve the problem?

**Comment by intrip** [comment]
Can you reproduce the issue with the following [reproduction script](https://gist.github.com/rafaelfranca/dd0a2d32397fd9c3a596a349e6601309) ?

**Comment by choncou** [comment]
Here's a reproduction script and results:
### Script
```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  # Activate the gem you are reporting the issue against.
  gem "activerecord", "6.1.0"
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
    t.integer :lockable_comments_count
    t.integer :comments_count
  end

  create_table :comments, force: true do |t|
    t.integer :post_id
  end

  create_table :lockable_comments, force: true do |t|
    t.integer :lock_version
    t.integer :post_id
  end
end

class Post < ActiveRecord::Base
  has_many :comments, dependent: :destroy
  has_many :lockable_comments, dependent: :destroy
end

class Comment < ActiveRecord::Base
  belongs_to :post, counter_cache: 'comments_count'
end

class LockableComment < ActiveRecord::Base
  belongs_to :post, counter_cache: 'lockable_comments_count'
end

class BugTest < Minitest::Test
  def test_regular_counter_cache
    post = Post.create!
    post.comments << Comment.create!
    post.comments << Comment.create!

    assert_equal 2, post.comments.count
    assert_equal 2, post.comments_count

    post.comments.last.destroy
    assert_equal 1, post.comments.count
    assert_equal 1, post.comments_count
  end

  # FAILS
  def test_counter_cache_on_locked_association
    post = Post.create!
    post.lockable_comments << LockableComment.create!
    post.lockable_comments << LockableComment.create!

    assert_equal 2, post.lockable_comments.count
    assert_equal 2, post.lockable_comments_count

    post.lockable_comments.last.destroy
    assert_equal 1, post.lockable_comments.count
    assert_equal 1, post.lockable_comments_count
  end
end

```

### Results
```
Finished in 0.020077s, 99.6165 runs/s, 398.4659 assertions/s.

  1) Failure:
BugTest#test_counter_cache_on_locked_association [ar.rb:77]:
Expected: 1
  Actual: 2

2 runs, 8 assertions, 1 failures, 0 errors, 0 skips
```

**Comment by choncou** [comment]
I've created a PR which I think might do the job here

**Comment by choncou** [comment]
Fixed by #42282
